README.md
# [PCem](https://pcem-emulator.co.uk/)
![C/C++ CI](https://github.com/sarah-walker-pcem/pcem/actions/workflows/c-cpp.yml/badge.svg)
## Download: [Windows](https://pcem-emulator.co.uk/files/PCemV17Win.zip)/[Linux](https://pcem-emulator.co.uk/files/PCemV17Linux.tar.gz)

Latest version: <b>v17</b> [Changelog](CHANGELOG.md)

PCem is licensed under GPL v2.0, see [COPYING](COPYING) for more details.

You can submit patches on our [forum](https://pcem-emulator.co.uk/phpBB3). Before you do, please note the [guidelines](https://pcem-emulator.co.uk/phpBB3/viewtopic.php?f=3&t=5) for submitting patches.

:exclamation: Note: <b>NO COPYRIGHTED ROM FILES ARE INCLUDED NOR WILL THEY BE. PLEASE DO NOT ASK FOR THEM.</b>

## BSD and Linux supplement (v17)

You will need the following libraries and buildtools (and their dependencies):
- SDL2
- Boost
- OpenAL
- CMake
- Ninja (Recommended, but you can use a Makefile generator if you prefer)

Depending on which GUI Library you are building, you may need Qt as well

Open a terminal window, navigate to the PCem directory then enter:
### Linux/BSD
```
cmake -G "Ninja" -DCMAKE_BUILD_TYPE=Release .
ninja
```

### Windows (MSYS2)
```
cmake -G "Ninja" -DMSYS=TRUE -DCMAKE_BUILD_TYPE=Release .
ninja
```

then `./src/pcem` to run.

The Linux/BSD versions store BIOS ROM images, configuration files, and other data in `~/.pcem`

configure options are :
```
  -DCMAKE_BUILD_TYPE=Release : Generate release build. Recommended for regular use.
  -DCMAKE_BUILD_TYPE=Debug   : Compile with debugging enabled.
  -DUSE_NETWORKING=OFF       : Build with networking support.
  -DUSE_PCAP_NETWORKING=ON   : Build with pcap networking support. (Needs USE_NETWORKING to compile) Requires libpcap.
  -DUSE_ALSA=OFF             : Build with support for MIDI output through ALSA. Requires libasound. (Linux Only)
  -DPLUGIN_ENGINE=OFF        : Build with plugin support. Builds libpcem-plugin-api and links PCem with it.
  -DPCEM_GUI_MODE=SDL        : What GUI to build with PCem. The following is supported
    * SDL       : SDL means only SDL mode, no configuration GUI will be included. You will launch a config with the command line
```

If you are using -DCMAKE_BUILD_TYPE=Debug, there are some more debug options you can enable if needed
```
  -DPCEM_SLIRP_DEBUG=ON           : Build PCem with SLIRP_DEBUG debug output
  -DPCEM_RECOMPILER_DEBUG=ON      : Build PCem with RECOMPILER_DEBUG debug output
  -DPCEM_NE2000_DEBUG=ON          : Build PCem with NE2000_DEBUG debug output
  -DPCEM_EMU8K_DEBUG_REGISTERS=ON : Build PCem with EMU8K_DEBUG_REGISTERS debug output
  -DPCEM_SB_DSP_RECORD_DEBUG=ON   : Build PCem with SB_DSP_RECORD_DEBUG debug output
  -DPCEM_MACH64_DEBUG=ON          : Build PCem with MACH64_DEBUG debug output
  -DPCEM_DEBUG_EXTRA=ON           : Build PCem with DEBUG_EXTRA debug output
```

If you are using -DCMAKE_BUILD_TYPE=RelWithDebInfo, there are additional options you can do
```
  -DPCEM_RELDEB_AS_RELEASE=ON     : Builds RelWithDebInfo with debugging logging enabled when this is off
```

They are some extra modules you can add if you build with `-DUSE_EXPERIMENTAL=ON`. These modules are untested.
incomplete, and may or may not be in a future build of PCem. We do not provide builds with these enabled as
well. It is also possible they may not even build.
```
  -DUSE_EXPERIMENTAL_PGC=ON       : Build PCem with Professional Graphics Controller support.
  -DUSE_EXPERIMENTAL_PRINTER=ON   : Build PCem with Printer support. Requires freetype.
``` 
The menu is a pop-up menu in the Linux/BSD port. Right-click on the main window when mouse is not
captured.

CD-ROM support currently only accesses `/dev/cdrom`. It has not been heavily tested.

## Links

### PCem emulates the following hardware (as of v17):

Hardware | Links
--- | ---
Systems | [8088](#8088-based)<br/>[8086](#8086-based)<br/>[286](#286-based)<br/>[386](#386-based)<br/>[486](#486-based)<br/>[Pentium](#pentium-based)<br/>[Super Socket 7](#super-socket-7-based)
Graphics | [Basic](#basic-cards)<br/>[Unaccelerated (S)VGA cards](#unaccelerated-svga-cards)<br/>[2D accelerated SVGA cards](#2d-accelerated-svga-cards)<br/>[3D accelerated SVGA cards](#3d-accelerated-svga-cards)<br/>[3D only cards](#3d-only-cards)
Sound | [Cards](#sound-cards)
HDD Controller | [Cards](#hdd-controller-cards)
Misc | [Cards](#misc-cards)

### [Software Tested](TESTED.md) (list)
- [DOS](TESTED.md#dos)<br/>
- [Windows](TESTED.md#windows)<br/>
- [Windows NT](TESTED.md#windows-nt)<br/>
- [OS/2](TESTED.md#os2)<br/>
- [Linux](TESTED.md#linux)<br/>
- [Applications](TESTED.md#applications)<br/>
- [Games](TESTED.md#games)<br/>
- [Demos](TESTED.md#demos)<br/>
- [Emulators](TESTED.md#emulators)<br/>

<hr>

## Systems

### 8088 based
Release | Machine | ROM file needed<br/>(within ./roms/ folder)
:-: | --- | ---
1981 | <b>IBM PC</b><br/>8088 at 4.77 MHz<br/>16KB - 640KB RAM (min. 64KB) | ibmpc/pc102782.bin<br/>ibmpc/basicc11.f6<br/>ibmpc/basicc11.f8<br/>ibmpc/basicc11.fa<br/>ibmpc/basicc11.fc
1983 | <b>Compaq Portable Plus</b><br/>8088 at 4.77 MHz<br/>128KB - 640KB RAM | compaq_pip/Compaq Portable Plus 100666-001 Rev C.bin
1983 | <b>IBM XT</b><br/>8088 at 4.77 MHz<br/>64KB - 640KB RAM | ibmxt/5000027.u19<br/>ibmxt/1501512.u18
1983 | <b>Leading Edge Model M</b><br/>8088 at 7.16 MHz<br/>128KB - 704KB RAM | leadingedge_modelm/Leading Edge - Model M - BIOS ROM - Version 4.71.bin
1984 | <b>IBM PCjr</b> <i>[[5]](#system-note-5)</i><br/>8088 at 4.77 MHz<br/>64KB - 640KB RAM (min. 128KB)<br/>Built-in 16 colour graphics<br/>3 voice sound<br/>Not generally PC compatible. | ibmpcjr/bios.rom
1984 | <b>Tandy 1000</b> <i>[[5]](#system-note-5)</i><br/>8088 at 4.77 MHz<br/>128KB - 640KB RAM<br/>Built-in 16 colour graphics<br/>3 voice sound | tandy/tandy1t1.020
1985 | <b>Commodore PC10</b><br/>8088 at 4.77 MHz<br/>640KB RAM CGA/Monochrome | cbm_pc10/cbm-pc10c-bios-v4.41-318085-08.bin
1985 | <b>NCR PC4i</b><br/>8088 at 4.77 MHz<br/>256KB - 640KB RAM | ncr_pc4i/NCR_PC4i_BIOSROM_1985.BIN
1986 | <b>DTK Clone XT</b><br/>8088 at 8/10 MHz<br/>64KB - 640KB RAM | dtk/dtk_erso_2.42_2764.bin
1986 | <b>Phoenix XT clone</b><br/>8088 at 8/10 MHz<br/>64KB - 640KB RAM | pxxt/000p001.bin
1987 | <b>Hyundai Super 16T</b><br/>8088/8087 at 4.77/8.0 MHz<br/>640KB RAM<br/>EGA 16 colour graphics<br/> | ./huyndaixte/super16te_v2.00id.bin
1987 | <b>Tandy 1000HX</b><br/>8088 at 7.16 MHz<br/>256KB - 640KB RAM<br/>Built-in 16 colour graphics<br/>3 voice sound<br/>Has DOS 2.11 in ROM | tandy1000hx/v020000.u12
1987 | <b>Thomson TO16 PC</b><br/>8088 at 9.54 MHz<br/>512KB - 640KB RAM | to16_pc/TO16_103.bin
1987 | <b>Toshiba T1000</b> <i>[[1]](#system-note-1)</i> <i>[[5]](#system-note-5)</i><br/>8088 at 4.77 MHz<br/>512KB - 1024KB RAM<br/>CGA on built-in LCD | t1000/t1000.rom<br/>t1000/t1000font.rom
1987 | <b>VTech Laser Turbo XT</b><br/>8088 at 10 MHz<br/>640KB RAM | ltxt/27c64.bin
1987 | <b>Zenith Data SupersPort</b><br/>8088 at 8 MHz<br/>128KB - 640KB RAM<br/>Built-in LCD video is not currently emulated | zdsupers/z184m v3.1d.10d
1988? | <b>&#169;Anonymous Generic Turbo XT BIOS</b><br/>8088 at 8+ MHz<br/>64KB - 640KB RAM | genxt/pcxt.rom
1988 | <b>Atari PC3</b><br/>8088 at 8 MHz<br/>640KB RAM | ataripc3/AWARD_ATARI_PC_BIOS_3.08.BIN
1988 | <b>Juko XT clone</b> | jukopc/000o001.bin
1988 | <b>Schneider Euro PC</b><br/>8088 at 9.54 MHz<br/>512KB - 640KB RAM | europc/50145<br/>europc/50146
1989 | <b>AMI XT clone</b><br/>8088 at 8+ MHz<br/>64KB - 640KB RAM | amixt/ami_8088_bios_31jan89.bin
1989 | <b>Hyundai Super 16TE</b><br/>8088/8087 at 10 MHz<br/>640KB RAM<br/>EGA 16 colour graphics<br/> | ./huyndaixte/super16te_v2.00id.bin
2015 | <b>Xi8088</b><br/>8088 at 4.77-13.33 MHz<br/>640KB RAM | xi8088/bios-xi8088.bin

### 8086 based
Release | Machine | ROM file needed<br/>(within ./roms/ folder)
:-: | --- | ---
1984 | <b>Compaq Deskpro</b><br/>8086 at 8 MHz<br/>128KB - 640KB RAM | deskpro/Compaq - BIOS - Revision J - 106265-002.bin
1984 | <b>Olivetti M24</b> <i>[[5]](#system-note-5)</i><br/>8086 at 8 MHz<br/>128KB - 640KB RAM<br/>Built-in enhanced CGA (supports 640x400x2) | olivetti_m24/olivetti_m24_version_1.43_low.bin<br/>olivetti_m24/olivetti_m24_version_1.43_high.bin
1986 | <b>Amstrad PC1512</b> <i>[[5]](#system-note-5)</i><br/>8086 at 8 MHz<br/>512KB - 640KB RAM<br/>Enhanced CGA (supports 640x200x16)<br/>Custom mouse port | pc1512/40043.v1<br/>pc1512/40044.v2<br/>pc1512/40078.ic127
1987 | <b>Amstrad PC1640</b> <i>[[5]](#system-note-5)</i><br/>8086 at 8 MHz<br/>640KB RAM<br/>Built-in Paradise EGA<br/>Custom mouse port | pc1640/40043.v3<br/>pc1640/40044.v3<br/>pc1640/40100
1987 | <b>Toshiba T1200</b> <i>[[1]](#system-note-1)</i> <i>[[5]](#system-note-5)</i><br/>8086 at 9.54 MHz<br/>1MB - 2MB RAM<br/>CGA on built-in LCD | t1200/t1200_019e.ic15.bin<br/>t1200/t1000font.rom
1988 | <b>Amstrad PPC512/640</b> <i>[[5]](#system-note-5)</i><br/>8086 at 8 MHz<br/>512KB - 640KB RAM<br/>Built-in CGA w/ plasma display | ppc512/40107.v2<br/>ppc512/40108.v2<br/>ppc512/40109.bin
1988 | <b>Sinclair PC200/Amstrad PC20</b> <i>[[5]](#system-note-5)</i><br/>8086 at 8 MHz<br/>512KB - 640KB RAM<br/>Built-in CGA (supports TV-out 50hz PAL) | pc200/pc20v2.0<br/>pc200/pc20v2.1<br/>pc200/40109.bin
1988 | <b>VTech Laser XT3</b><br/>8086 at 10 MHz<br/>512KB - 1152KB RAM | lxt3/27c64d.bin
1989 | <b>Amstrad PC2086</b> <i>[[5]](#system-note-5)</i><br/>8086 at 8 MHz<br/>640KB RAM<br/>Built-in VGA | pc2086/40179.ic129<br/>pc2086/40180.ic132<br/>pc2086/40186.ic171
1989 | <b>Tandy 1000SL/2</b> <i>[[5]](#system-note-5)</i><br/>8086 at 9.54 MHz<br/>512KB - 768KB RAM<br/>Built-in 16 colour graphics<br/>4 voice sound | tandy1000sl2/8079047.hu1<br/>tandy1000sl2/8079048.hu2
1990 | <b>Amstrad PC3086</b> <i>[[5]](#system-note-5)</i><br/>8086 at 8 MHz<br/>640KB RAM<br/>Built-in VGA | pc3086/fc00.bin<br/>pc3086/c000.bin
1991 | <b>Amstrad PC5086</b><br/>8086 at 8 MHz<br/>640KB RAM | pc5086/sys_rom.bin

### 286 based
Release | Machine | ROM file needed<br/>(within ./roms/ folder)
:-: | --- | ---
1984 | <b>IBM AT</b><br/>286 at 6 or 8 MHz<br/>256KB - 16MB RAM | ibmat/at111585.0<br/>ibmat/at111585.1
1986 | <b>Compaq Portable II</b><br/>286 at 8 MHz<br/>256KB - 15MB RAM | compaq_pii/109739-001.rom<br/>compaq_pii/109740-001.rom
1986 | <b>IBM XT Model 286</b><br/>286 at 6 MHz<br/>256KB - 16MB RAM | ibmxt286/BIOS_5162_21APR86_U34_78X7460_27256.BIN<br/>ibmxt286/BIOS_5162_21APR86_U35_78X7461_27256.BIN
1986 | <b>Toshiba T3100e</b> <i>[[1]](#system-note-1)</i> <i>[[5]](#system-note-5)</i><br/>286 at 12 MHz<br/>1MB - 5MB RAM<br/>CGA on gas-plasma display | t3100e/t3100e_font.bin<br/>t3100e/t3100e.rom
1987 | <b>IBM PS/2 Model 50</b> <i>[[5]](#system-note-5)</i><br/>286 at 10 MHz<br/>1MB - 16MB RAM<br/>Built-in VGA<br/>MCA bus | i8550021/90x7420.zm13<br/>i8550021/90x7423.zm14<br/>i8550021/90x7426.zm16<br/>i8550021/90x7429.zm18
1988 | <b>Bull Micral 45</b><br/>286 at 12 MHz<br/>1MB - 6MB RAM | bull_micral_45/even.fil<br/>bull_micral_45/odd.fil
1988 | <b>Commodore PC30-III</b><br/>286 at 12 MHz<br/>512KB - 16MB RAM | cmdpc30/commodore pc 30 iii even.bin<br/>cmdpc30/commodore pc 30 iii odd.bin
1988 | <b>IBM PS/2 Model 30-286</b><br/>286 at 10 MHz<br/>1MB - 16MB RAM<br/>Built-in VGA<br/>MCA bus | ibmps2_m30_286/33f5381a.bin
1989 | <b>Epson PC AX</b><br/>286<br/>256KB - 16MB RAM | epson_pcax/EVAX<br/>epson_pcax/ODAX
1989 | <b>Epson PC AX2e</b><br/>286 at 12 MHz<br/>256KB - 16MB RAM | epson_pcax2e/EVAX<br/>epson_pcax2e/ODAX
1990 | <b>AMI 286 clone</b><br/>286 at 8+ MHz<br/>512KB - 16MB RAM | ami286/amic206.bin
1990 | <b>Award 286 clone</b><br/>286 at 8+ MHz<br/>512KB - 16MB RAM | award286/award.bin
1990 | <b>Dell System 200</b><br/>286 at 12 MHz<br/>640KB - 16MB RAM | dells200/dell0.bin<br/>dells200/dell1.bin
1990 | <b>IBM PS/1 Model 2011</b> <i>[[5]](#system-note-5)</i><br/>286 at 10 MHz<br/>512KB - 16MB RAM<br/>Built-in VGA<br/>DOS 4.01 + GUI menu system in ROM | ibmps1/f80000.bin
? | <b>Goldstar GDC-212M</b><br/>286 at 12 MHz<br/>512KB - 4MB RAM | gdc212m/gdc212m_72h.bin
? | <b>GW-286CT GEAR</b><br/>286 at 8+ MHz<br/>512KB - 16MB RAM | gw286ct/2ctc001.bin
? | <b>Hyundai Super-286TR</b><br/>286 at 12 MHz<br/>1MB - 4MB RAM | super286tr/award.bin
? | <b>Samsung SPC-4200P</b><br/>286 at 12 MHz<br/>512KB - 2MB RAM | spc4200p/u8.01

debian/changelog
pcem (17-0utopia2) unstable; urgency=medium

  * Rebuild against libwxgtk3.2-dev

 -- James Lu <james@overdrivenetworks.com>  Sun, 26 Mar 2023 00:17:26 -0700

pcem (17-0utopia1) unstable; urgency=medium

  * Utopia repository upload.
  * Initial release.

 -- James Lu <james@overdrivenetworks.com>  Wed, 22 Jun 2022 23:45:27 -0700


src/sdl/sdl2.c
#include <SDL2/SDL.h>

#ifdef __WINDOWS__
#define BITMAP WINDOWS_BITMAP
#undef UNICODE
#include <windows.h>
#include <windowsx.h>
#undef BITMAP
#endif

#include <string.h>
#include <stdio.h>
#include <stdarg.h>
#include <stdlib.h>
#include <math.h>
#include "ibm.h"
#include "device.h"
#include "cassette.h"
#include "cdrom-ioctl.h"
#include "cdrom-image.h"
#include "config.h"
#include "video.h"
#include "cpu.h"
#include "ide.h"
#include "hdd.h"
#include "model.h"
#include "mouse.h"
#include "nvr.h"
#include "lpt.h"
#include "plat-joystick.h"
#include "plat-midi.h"
#include "scsi_zip.h"
#include "sound.h"
#include "thread.h"
#include "disc.h"
#include "disc_img.h"
#include "mem.h"
#include "paths.h"
#include "nethandler.h"

#include "sdl2-video.h"
#include "sdl2-display.h"

#include "plugin.h"
#include "pic.h"

#if __APPLE__
#define pause __pause
#include <util.h>
#include <fcntl.h>
#include <unistd.h>
#undef pause
#include <sys/types.h>
#include <sys/stat.h>
#endif

typedef enum
{
	EMULATION_STOPPED,
	EMULATION_PAUSED,
	EMULATION_RUNNING
} emulation_state_t;

#ifndef TRUE
#define TRUE 1
#endif

#ifndef FALSE
#define FALSE 0
#endif

extern void creatediscimage_open(void *hwnd);

#define IDM_CDROM_REAL 1500

#define MIN_SND_BUF 50

uint64_t timer_freq;

int gfx_present[GFX_MAX];

SDL_mutex *ghMutex;
SDL_mutex *mainMutex;
SDL_cond *mainCond;

SDL_Thread *mainthreadh = NULL;

SDL_TimerID onesectimer;

int running = 0;

int drawits = 0;

int romspresent[ROM_MAX];
int quited = 0;

SDL_Rect oldclip;

void *ghwnd = 0;

void *menu;

emulation_state_t emulation_state = EMULATION_STOPPED;
int pause = 0;

int window_doreset = 0;
int window_dosetresize = 0;
int renderer_doreset = 0;
int window_dofullscreen = 0;
int window_dowindowed = 0;
int window_doremember = 0;
int window_doinputgrab = 0;
int window_doinputrelease = 0;
int window_dotogglefullscreen = 0;

int video_scale = 1;

int video_width = 640;
int video_height = 480;

char menuitem[60];

extern int config_selection_open(void *hwnd, int inited);
extern int shader_manager_open(void *hwnd);

extern void sdl_set_window_title(const char *title);

extern float gl3_shader_refresh_rate;
extern float gl3_input_scale;
extern int gl3_input_stretch;
extern char gl3_shader_file[20][512];

char screenshot_format[10];
int screenshot_flash = 1;
int take_screenshot = 0;

void updatewindowsize(int x, int y)
{
	if (video_width == x && video_height == y)
	{
		return;
	}
	video_width = x;
	video_height = y;

	display_resize(x, y);
}

unsigned int get_ticks()
{
	return SDL_GetTicks();
}

void delay_ms(unsigned int ms)
{
	SDL_Delay(ms);
}

void startblit()
{
	SDL_LockMutex(ghMutex);
}

void endblit()
{
	SDL_UnlockMutex(ghMutex);
}

void enter_fullscreen()
{
	window_dofullscreen = window_doinputgrab = 1;
}

void leave_fullscreen()
{
	window_dowindowed = window_doinputrelease = 1;
}

void toggle_fullscreen()
{
	window_dotogglefullscreen = 1;
}

uint64_t main_time;

int mainthread(void *param)
{
	SDL_SetThreadPriority(SDL_THREAD_PRIORITY_HIGH);

	int frames = 0;
	uint32_t old_time, new_time;

	drawits = 0;
	old_time = SDL_GetTicks();
	running = 1;
	while (running)
	{
		new_time = SDL_GetTicks();
		drawits += new_time - old_time;
		old_time = new_time;

		if (drawits > 0 && !pause)
		{
			uint64_t start_time = timer_read();
			uint64_t end_time;
			drawits -= 10;
			if (drawits > 50)
				drawits = 0;
			runpc();
			frames++;
			if (frames >= 200 && nvr_dosave)
			{
				frames = 0;
				nvr_dosave = 0;
				savenvr();
			}
			end_time = timer_read();
			main_time += end_time - start_time;
		}
		else
			SDL_Delay(1);
	}

	SDL_LockMutex(mainMutex);
	SDL_CondSignal(mainCond);
	SDL_UnlockMutex(mainMutex);

	return TRUE;
}

void stop_emulation_now(void)
{
	/*Deduct a sufficiently large number of cycles that no instructions will
	  run before the main thread is terminated*/
	cycles -= 99999999;
}

void set_window_title(const char *s)
{
	sdl_set_window_title(s);
}

float flash_func(float x)
{
	return 1 - pow(x, 4);
}

float flash_failed_func(float x)
{
	return fabs(sin(x * 3.1415926 * 2));
}

void screenshot_taken(unsigned char *rgb, int width, int height)
{
// TODO: Use non wxWidgets API
#if 0
	char name[512];
	char date[128];
	strcpy(name, "Screenshot from ");
	wx_date_format(date, "%Y-%m-%d %H-%M-%S");
	strcat(name, date);
	if (wx_image_save(screenshots_path, name, screenshot_format, rgb, width, height, 0))
	{
		pclog("Screenshot saved\n");
		if (screenshot_flash)
			color_flash(flash_func, 500, 0xff, 0xff, 0xff, 0xff);
	}
	else
	{
		pclog("Screenshot was not saved\n");
		if (screenshot_flash)
			color_flash(flash_failed_func, 500, 0xff, 0, 0, 0xff);
	}
#endif
}

uint64_t timer_read()
{
	return SDL_GetPerformanceCounter();
}

Uint32 timer_onesec(Uint32 interval, void *param)
{
	onesec();
	return interval;
}

void sdl_loadconfig()
{
	vid_resize = config_get_int(CFG_MACHINE, NULL, "vid_resize", 0);
	video_fullscreen_scale = config_get_int(CFG_MACHINE, NULL, "video_fullscreen_scale", 0);
	video_fullscreen_first = config_get_int(CFG_MACHINE, NULL, "video_fullscreen_first", 1);

	strcpy(screenshot_format, config_get_string(CFG_MACHINE, "SDL2", "screenshot_format", 0));
	screenshot_flash = config_get_int(CFG_MACHINE, "SDL2", "screenshot_flash", 1);

	custom_resolution_width = config_get_int(CFG_MACHINE, "SDL2", "custom_width", custom_resolution_width);
	custom_resolution_height = config_get_int(CFG_MACHINE, "SDL2", "custom_height", custom_resolution_height);

	video_fullscreen = config_get_int(CFG_MACHINE, "SDL2", "fullscreen", video_fullscreen);
	video_fullscreen_mode = config_get_int(CFG_MACHINE, "SDL2", "fullscreen_mode", video_fullscreen_mode);
	video_scale = config_get_int(CFG_MACHINE, "SDL2", "scale", video_scale);
	video_scale_mode = config_get_int(CFG_MACHINE, "SDL2", "scale_mode", video_scale_mode);
	video_vsync = config_get_int(CFG_MACHINE, "SDL2", "vsync", video_vsync);
	video_focus_dim = config_get_int(CFG_MACHINE, "SDL2", "focus_dim", video_focus_dim);
	video_alternative_update_lock = config_get_int(CFG_MACHINE, "SDL2", "alternative_update_lock", video_alternative_update_lock);
	requested_render_driver = sdl_get_render_driver_by_name(config_get_string(CFG_MACHINE, "SDL2", "render_driver", ""), RENDERER_SOFTWARE);

#if 0
	gl3_input_scale = config_get_float(CFG_MACHINE, "GL3", "input_scale", gl3_input_scale);
	gl3_input_stretch = config_get_int(CFG_MACHINE, "GL3", "input_stretch", gl3_input_stretch);
	gl3_shader_refresh_rate = config_get_float(CFG_MACHINE, "GL3", "shader_refresh_rate", gl3_shader_refresh_rate);

	memset(&gl3_shader_file, 0, sizeof(gl3_shader_file));
	int num_shaders = config_get_int(CFG_MACHINE, "GL3 Shaders", "shaders", 0);
	char s[20];
	int i;
	for (i = 0; i < num_shaders; ++i)
	{
		sprintf(s, "shader%d", i);
		strncpy(gl3_shader_file[i], config_get_string(CFG_MACHINE, "GL3 Shaders", s, ""), 511);
		gl3_shader_file[i][511] = 0;
	}
#endif
}

void sdl_saveconfig()
{
	config_set_int(CFG_MACHINE, NULL, "vid_resize", vid_resize);
	config_set_int(CFG_MACHINE, NULL, "video_fullscreen_scale", video_fullscreen_scale);
	config_set_int(CFG_MACHINE, NULL, "video_fullscreen_first", video_fullscreen_first);

	config_set_string(CFG_MACHINE, "SDL2", "screenshot_format", screenshot_format);
	config_set_int(CFG_MACHINE, "SDL2", "screenshot_flash", screenshot_flash);

	config_set_int(CFG_MACHINE, "SDL2", "custom_width", custom_resolution_width);
	config_set_int(CFG_MACHINE, "SDL2", "custom_height", custom_resolution_height);

	config_set_int(CFG_MACHINE, "SDL2", "fullscreen", video_fullscreen);
	config_set_int(CFG_MACHINE, "SDL2", "fullscreen_mode", video_fullscreen_mode);
	config_set_int(CFG_MACHINE, "SDL2", "scale", video_scale);
	config_set_int(CFG_MACHINE, "SDL2", "scale_mode", video_scale_mode);
	config_set_int(CFG_MACHINE, "SDL2", "vsync", video_vsync);
	config_set_int(CFG_MACHINE, "SDL2", "focus_dim", video_focus_dim);
	config_set_int(CFG_MACHINE, "SDL2", "alternative_update_lock", video_alternative_update_lock);
	config_set_string(CFG_MACHINE, "SDL2", "render_driver", (char *)requested_render_driver.sdl_id);

#if 0
	config_set_float(CFG_MACHINE, "GL3", "input_scale", gl3_input_scale);
	config_set_int(CFG_MACHINE, "GL3", "input_stretch", gl3_input_stretch);
	config_set_float(CFG_MACHINE, "GL3", "shader_refresh_rate", gl3_shader_refresh_rate);

	char s[20];
	int i;
	for (i = 0; i < 20; ++i)
	{
		sprintf(s, "shader%d", i);
		if (strlen(gl3_shader_file[i]))
			config_set_string(CFG_MACHINE, "GL3 Shaders", s, gl3_shader_file[i]);
		else
			break;
	}
	config_set_int(CFG_MACHINE, "GL3 Shaders", "shaders", i);
#endif
}

void sdl_onconfigloaded()
{
	/* create directories */
	if (!dir_exists(configs_path))
		dir_create(configs_path);
	if (!dir_exists(nvr_path))
		dir_create(nvr_path);
	if (!dir_exists(logs_path))
		dir_create(logs_path);
	if (!dir_exists(screenshots_path))
		dir_create(screenshots_path);
}

int pc_main(int argc, char **argv)
{
	// Expose some functions to libpcem-plugin-api without moving them over to
	// the plugin api proper
	_savenvr = savenvr;
	_dumppic = dumppic;
	_dumpregs = dumpregs;
	_sound_speed_changed = sound_speed_changed;

	paths_init();

	init_plugin_engine();
	model_init_builtin();
	video_init_builtin();
	lpt_init_builtin();
	sound_init_builtin();
	hdd_controller_init_builtin();
#ifdef USE_NETWORKING
	network_card_init_builtin();
#endif

	add_config_callback(sdl_loadconfig, sdl_saveconfig, sdl_onconfigloaded);

	initpc(argc, argv);
	resetpchard();

	sound_init();

#ifndef __APPLE__
	display_init();
#endif
	sdl_video_init();
	joystick_init();

	start_emulation(NULL);

	return TRUE;
}

int resume_emulation()
{
	if (emulation_state == EMULATION_PAUSED)
	{
		emulation_state = EMULATION_RUNNING;
		pause = 0;
		return TRUE;
	}
	return FALSE;
}

int start_emulation(void *params)
{
	if (resume_emulation())
		return TRUE;
	int c;
	pclog("Starting emulation...\n");
	loadconfig(NULL);

	emulation_state = EMULATION_RUNNING;
	pause = 0;

	ghMutex = SDL_CreateMutex();
	mainMutex = SDL_CreateMutex();
	mainCond = SDL_CreateCond();

	if (!loadbios())
	{
		if (romset != -1)
			// wx_messagebox(ghwnd,
			//			  "Configured romset not available.\nDefaulting to available romset.",
			//			  "PCem error", WX_MB_OK);
			pclog("Error: Configured romset not available.\nDefaulting to available romset.");
		for (c = 0; c < ROM_MAX; c++)
		{
			if (romspresent[c])
			{
				romset = c;
				model = model_getmodel(romset);
				break;
			}
		}
	}

	if (!video_card_available(video_old_to_new(gfxcard)))
	{
		if (romset != -1)
			// wx_messagebox(ghwnd,
			//			  "Configured video BIOS not available.\nDefaulting to available romset.",
			//			  "PCem error", WX_MB_OK);
			pclog("Error: Configured video BIOS not available.\nDefaulting to available romset.");
		for (c = GFX_MAX - 1; c >= 0; c--)
		{
			if (gfx_present[c])
			{
				gfxcard = c;
				break;
			}
		}
	}

	loadbios();
	resetpchard();
	midi_init();

	display_start(params);

	mainthreadh = SDL_CreateThread(mainthread, "Main Thread", NULL);

	onesectimer = SDL_AddTimer(1000, timer_onesec, NULL);

	updatewindowsize(640, 480);

	timer_freq = SDL_GetPerformanceFrequency();

	// if (show_machine_on_start)
	//	wx_show_status(ghwnd);
	while (emulation_state == EMULATION_RUNNING)
		;

	return TRUE;
}

int pause_emulation()
{
	pclog("Emulation paused.\n");
	emulation_state = EMULATION_PAUSED;
	pause = 1;
	return TRUE;
}

int stop_emulation()
{
	emulation_state = EMULATION_STOPPED;
	pclog("Stopping emulation...\n");
	SDL_LockMutex(mainMutex);
	running = 0;
	SDL_CondWaitTimeout(mainCond, mainMutex, 10 * 1000);
	SDL_UnlockMutex(mainMutex);

	SDL_DestroyCond(mainCond);
	SDL_DestroyMutex(mainMutex);

	startblit();
	display_stop();

#if SDL_VERSION_ATLEAST(2, 0, 2)
	SDL_DetachThread(mainthreadh);
#endif
	mainthreadh = NULL;
	SDL_RemoveTimer(onesectimer);
	savenvr();
	saveconfig(NULL);

	endblit();
	SDL_DestroyMutex(ghMutex);

	device_close_all();
	midi_close();

	pclog("Emulation stopped.\n");

	return TRUE;
}

void reset_emulation()
{
	pause_emulation();
	SDL_Delay(100);
	resetpchard();
	resume_emulation();
}

int wx_stop()
{
	pclog("Shutting down...\n");
	closepc();
	display_close();
	sdl_video_close();

	printf("Shut down successfully!\n");
	return TRUE;
}
